rpctest/rpc_harness.go

```diff
@@ -0,0 +1,395 @@
+}
+
+// CreateTransaction returns a fully signed transaction paying to the specified
+// outputs while observing the desired fee rate. The passed fee rate should be
+// expressed in satoshis-per-byte.
+//
+// This function is safe for concurrent access.
+func (h *Harness) CreateTransaction(targetOutputs []*wire.TxOut,
+	feeRate btcutil.Amount) (*wire.MsgTx, error) {
+
+	return h.wallet.CreateTransaction(targetOutputs, feeRate)
+}
+
+// RPCConfig returns the harnesses current rpc configuration. This allows other
+// potential RPC clients created within tests to connect to a given test
+// harness instance.
+func (h *Harness) RPCConfig() btcrpcclient.ConnConfig {
+	return h.node.config.rpcConnConfig()
+}
+
+// GenerateAndSubmitBlock creates a block whose contents include the passed
+// transactions and submits it to the running simnet node. For generating
+// blocks with only a coinbase tx, callers can simply pass nil instead of
+// transactions to be mined. Additionally, a custom block version can be set by
+// the caller. A blockVersion of -1 indicates that the current default block
+// version should be used.
+//
+// This function is safe for concurrent access.
+func (h *Harness) GenerateAndSubmitBlock(txns []*btcutil.Tx, blockVersion int32,
+	blockTime *time.Time) (*btcutil.Block, error) {
+
+	h.Lock()
+	defer h.Unlock()
+
+	if blockVersion == -1 {
+		blockVersion = wire.BlockVersion
+	}
+
+	prevBlockHash, prevBlockHeight, err := h.Node.GetBestBlock()
+	if err != nil {
+		return nil, err
+	}
+	prevBlock, err := h.Node.GetBlock(prevBlockHash)
+	if err != nil {
+		return nil, err
+	}
+	prevBlock.SetHeight(prevBlockHeight)
+
+	// Create a new block including the specified transactions
+	newBlock, err := createBlock(prevBlock, txns, blockVersion,
+		blockTime, h.wallet.coinbaseAddr, h.ActiveNet)
+	if err != nil {
+		return nil, err
+	}
+
+	// Submit the block to the simnet node.
+	if err := h.Node.SubmitBlock(newBlock, nil); err != nil {
+		return nil, err
+	}
+
+	return newBlock, nil
+}
+
+// generateListeningAddresses returns two strings representing listening
+// addresses designated for the current rpc test. If there haven't been any
+// test instances created, the default ports are used. Otherwise, in order to
+// support multiple test nodes running at once, the p2p and rpc port are
+// incremented after each initialization.
+func generateListeningAddresses() (string, string) {
+	var p2p, rpc string
+	localhost := "127.0.0.1"
+
+	if numTestInstances == 0 {
+		p2p = net.JoinHostPort(localhost, strconv.Itoa(defaultP2pPort))
+		rpc = net.JoinHostPort(localhost, strconv.Itoa(defaultRPCPort))
+	} else {
+		p2p = net.JoinHostPort(localhost,
+			strconv.Itoa(defaultP2pPort+(2*numTestInstances)))
+		rpc = net.JoinHostPort(localhost,
+			strconv.Itoa(defaultRPCPort+(2*numTestInstances)))
+	}
+
+	return p2p, rpc
+}
+
+func init() {
+	// Create the testInstances map once the package has been imported.
+	testInstances = make(map[string]*Harness)
```

Comment:
Very true, also saves a few lines to boot. Updated to create it in-line with the declaration.